bundles/org.openhab.ui/web/src/assets/definitions/blockly/blocks-items.js

```diff
@@ -83,29 +83,84 @@ export default function (f7) {
+  * - category: String
+  * - tags: Array
+  * - groups: Array
+  * - type: String
+  * Blockly part
+  */
   Blockly.Blocks['oh_getitem_attribute'] = {
     init: function () {
+      let thisBlock = this
+      let dropdown = new Blockly.FieldDropdown(
+        [['name', 'Name'], ['label', 'Label'], ['state', 'State'], ['category', 'Category'], ['tags', 'Tags'], ['groups', 'GroupNames'], ['type', 'Type']],
+        function (newMode) {
+          thisBlock.updateType_(newMode)
+        })
       this.appendValueInput('item')
         .appendField('get ')
-        .appendField(new Blockly.FieldDropdown([['name', 'Name'], ['label', 'Label'], ['state', 'State'], ['category', 'Category'], ['tags', 'Tags'], ['groups', 'GroupNames'], ['type', 'Type']]), 'attributeName')
+        .appendField(dropdown, 'attributeName')
         .appendField('of item')
         .setCheck('oh_itemtype')
       this.setInputsInline(false)
-      this.setOutput(true, 'any')
+      this.setOutput(true, 'String')
       this.setColour(0)
-      this.setTooltip('Retrieve a specific attribute from the item')
+      this.setTooltip('Retrieve a specific attribute from the item. Note that groups and tags return a list and should be used with the loops-block \'for each item ... in list\'. ')
       this.setHelpUrl('https://www.openhab.org/docs/configuration/items.html')
+    },
+    /**
+    * Modify this block to have the correct output type based on the attribute.
+    */
+    updateType_: function (newAttributeName) {
+      let attributeName = this.getFieldValue('attributeName')
+      if (newAttributeName === 'Tags' || newAttributeName === 'GroupNames') {
+        this.outputConnection.setCheck('Array')
+      } else {
+        this.outputConnection.setCheck('String')
+      }
+    },
+    /**
+    * Create XML to represent the input and output types.
+    * @return {!Element} XML storage element.
+    * @this {Blockly.Block}
+    */
+    mutationToDom: function () {
+      let container = Blockly.utils.xml.createElement('mutation')
+      container.setAttribute('attributeName', this.getFieldValue('attributeName'))
+      return container
+    },
+    /**
+    * Parse XML to restore the input and output types.
+    * @param {!Element} xmlElement XML storage element.
+    * @this {Blockly.Block}
+    */
+    domToMutation: function (xmlElement) {
+      this.updateType_(xmlElement.getAttribute('attributeName'))
     }
   }
 
+  /*
+  * Provides all attributes from an item
+  * Code part
+  */
   Blockly.JavaScript['oh_getitem_attribute'] = function (block) {
     const theItem = Blockly.JavaScript.valueToCode(block, 'item', Blockly.JavaScript.ORDER_ATOMIC)
     const attributeName = block.getFieldValue('attributeName')
-    let code = `${theItem}.get${attributeName}()`
-    return [code, 0]
+    let code = `(${theItem} instanceof org.openhab.core.items.GenericItem) ? ${theItem}.get${attributeName}() : itemRegistry.getItem(${theItem}).get${attributeName}()`
```

Comment:
I'm really not sure we have to make the code more complex to handle this.

In the case of the _oh_getitem_state_ block, if we mistakenly supply an Item instance instead of a string, we get an exception that's actually confusing (the item exists):
![image](https://user-images.githubusercontent.com/2004147/147473679-c8b8eeee-4916-4714-bf82-1eda5fd8da6e.png)
```js
var i;


i = itemRegistry.getItem('Scene_General');
print(itemRegistry.getItem(i).getState());
```
> 13:53:26.751 [WARN ] [re.automation.internal.RuleEngineImpl] - Fail to execute action: script
java.lang.RuntimeException: org.openhab.core.items.ItemNotFoundException: Item 'Scene_General (Type=NumberItem, State=1, Label=Scene General, Category=house, Tags=[Status])' could not be found in the item registry
       at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:531) ~[jdk.scripting.nashorn:?]
        at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:456) ~[jdk.scripting.nashorn:?]
...


however if we supply something of a wrong type to the _oh_getitem_attribute_ block we get an error that's much easier to understand:
![image](https://user-images.githubusercontent.com/2004147/147474010-8e93587b-cb20-4432-a7d8-1912a8253be6.png)

```js
var n;


n = 'Scene_General';
print(n.getLabel());
```
> 13:57:31.201 [ERROR] [.internal.handler.ScriptActionHandler] - Script execution of rule with UID 'blockly_new' failed: TypeError: n.getLabel is not a function in <eval> at line number 5

That's more of a general problem the user have to be aware of (some blocks accept inputs of certain types, as suggested by the shadow blocks), and if you trick your way into supplying wrong types using variables you accept the risk of running into errors. It's not linked to OH blocks, for example:

![image](https://user-images.githubusercontent.com/2004147/147474420-00f24d70-115e-4679-95ea-d7bf65fb9926.png)

```js
var n, i;


n = 123;
print((n.join(',')));
```
> 14:05:07.345 [ERROR] [.internal.handler.ScriptActionHandler] - Script execution of rule with UID 'blockly_new' failed: TypeError: n.join is not a function in <eval> at line number 5

So I would leave it as-is in this case.

Reply:
> I'm really not sure we have to make the code more complex to handle this.

The case that came up that prompted this (at my suggestion) was when looping though the list of a Group's members one gets the Item Object. However, "get item state" expects an Item's name. 

I think it's somewhat unreasonable for the user to have to know and understand the difference in this one use case, especially when the same "get item state" block can be made to support both the Item's name and an actual Item Object transparently. 

It's the difference between:

![image](https://user-images.githubusercontent.com/6385096/147494648-264a6d0e-7ddb-4833-84ef-cd9e444ee124.png)

which intuitively seems like it should work but will fail and forcing the users to use

![image](https://user-images.githubusercontent.com/6385096/147494698-33c0f7eb-30a5-4273-b27e-f48ad527dbec.png)
which clearly works but is anything but intuitive. Why do I need to save the Item's state to a variable first? Why can't I just use "get state of item" directly? 

That is the specific problem being addressed here. I don't think, in this case at least, the users should have to know and deal with this type error so I'd say that merely changing the error so it's more meaningful is suboptimal. It would be more user friendly to just handle both cases (i.e. item name or item instance) transparently. 

As an aside, the number one problem I need to help new users with on the forum, regardless of the rules language chosen, are type problems. Anywhere we can avoid or sidestep a type errors for the users will be a big win. This is pretty firmly in the type error category.

I know we can't get away from all type errors, but this appears to be case where we can.